runtime/compiler/optimizer/OSRGuardInsertion.cpp

```diff
@@ -30,63 +30,205 @@
+         return true;
+      }
 
-   // Currently, OSR guard insertion is only needed when in the OSR HCR mode
-   if (disableOSRGuards || !comp()->isOSRTransitionTarget(TR::postExecutionOSR))
+   return true;
+   }
+
+static bool hasYieldPointCannotOSR(TR::Compilation *comp)
+   {
+   for (TR::TreeTop *treeTop = comp->getStartTree(); treeTop; treeTop = treeTop->getNextTreeTop())
       {
-      if (trace())
-         traceMsg(comp(), "HCR Guards are only removed when in OSR HCR mode - skipping\n");
-      return 0;
+      TR::Node *ttNode = treeTop->getNode();
+
+      if (comp->isPotentialOSRPoint(ttNode) &&
+          !comp->isPotentialOSRPointWithSupport(treeTop))
+         {
+         traceMsg(comp, "Found first bad yield point at n%dn\n", ttNode->getGlobalIndex());
+         return true;
+         }
       }
 
-   // Even under NextGenHCR, OSR may have been disabled for this compilation at runtime
-   if (!comp()->supportsInduceOSR())
+   return false;
+   }
+
+void TR_OSRGuardInsertion::cleanUpPotentialOSRPointHelperCalls()
+   {
+   for (TR::TreeTop *treeTop = comp()->getStartTree(); treeTop; treeTop = treeTop->getNextTreeTop())
       {
-      if (trace())
-         traceMsg(comp(), "OSR induce is not supported during this compilation - skipping\n");
-      return 0;
+      TR::Node *ttNode = treeTop->getNode();
+      if (ttNode->getNumChildren() != 1)
+         continue;
+
+      TR::Node *node = ttNode->getFirstChild();
+      if (node->isPotentialOSRPointHelperCall())
+         {
+         TR::TreeTop* prevTreeTop = treeTop->getPrevTreeTop();
+         TR::TransformUtil::removeTree(comp(), treeTop);
+         treeTop = prevTreeTop;
+         }
       }
+   }
 
-   // Detect if there are no HCR guards
-   TR::list<TR_VirtualGuard*> &virtualGuards = comp()->getVirtualGuards();
-   auto guard = virtualGuards.begin();
-   for (; guard != virtualGuards.end(); ++guard)
+void TR_OSRGuardInsertion::removeRedundantPotentialOSRPointHelperCalls(TR_HCRGuardAnalysis* guardAnalysis)
+   {
+   bool protectedByOSRPoints = false;
+   TR::NodeChecklist visited(comp());
+
+   for (TR::TreeTop *treeTop = comp()->getStartTree();
+        treeTop;
+        treeTop = treeTop->getNextRealTreeTop())
       {
-      if ((*guard)->getKind() == TR_HCRGuard || (*guard)->mergedWithHCRGuard())
-         break;
+      TR::Node *ttNode = treeTop->getNode();
+
+      if (ttNode->getOpCodeValue() == TR::BBStart)
+         {
+         TR::Block* block = treeTop->getEnclosingBlock();
+         protectedByOSRPoints = !guardAnalysis || guardAnalysis->_blockAnalysisInfo[block->getNumber()]->isEmpty();
+         continue;
+         }
+
+      TR::Node *osrNode = NULL;
+      if (comp()->isPotentialOSRPoint(ttNode, &osrNode))
+         {
+         if (visited.contains(osrNode))
+            continue;
+
+         if (protectedByOSRPoints &&
+             osrNode->isPotentialOSRPointHelperCall())
+            {
+            if (trace())
```

Comment:
Is this optional? If so it should be a perform transformation. If not this is probably a dumpOptDetails kind of thing.